### PR TITLE
doc: Update recommended way for installing cilium on AKS

### PR DESCRIPTION
**Pull request by tamilmani1989**
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Related: #27289

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

**Comment by gandro** [comment]
I know this has been discussed before, but we have users (e.g. @hemanthmalla) who are running Azure IPAM on Non-AKS. My understanding is that that usecase is not covered by either BYOCNI or Azure CNI Powered by Cilium, and the Azure IPAM is only legacy for AKS users, but not users who run self-managed K8s clusters on Azure. What's the plan there? 

I know that @hemanthmalla has expressed interest in the past for maintaining Azure IPAM for non-AKS use-cases.

**Comment by hemanthmalla** [comment]
@gandro Thanks for bringing this up. Yes, we use Azure IPAM mode in self managed kubernetes clusters on Azure. 
This came up a few times in community meetings, but I never really fully understood the rationale for this change.
@tamilmani1989 do you mind updating the PR to include more details about why we want to change this to legacy ?

**Comment by tamilmani1989** [comment]
@hemanthmalla I'm not aware of exact reason probably @joestringer would know but its already documented `legacy` [here](https://docs.cilium.io/en/stable/gettingstarted/k8s-install-default/#install-cilium ) and wanted to keep it consistent across all pages.

```
however integration with the Azure stack via the [Azure IPAM](https://docs.cilium.io/en/stable/network/concepts/ipam/azure/#ipam-azure) is not available and will only work with clusters not using BYOCNI. While still maintained for now, this mode is considered legacy.
```

**Comment by tamilmani1989** [comment]
Not sure about self managed clusters, but atleast for AKS, BYOCNI or Azure CNI Powered by Cilium is supported and recommended way to run cilium

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
(Disregarding my review, I initially looked at just the content and only later saw the discussion topics)

Broadly I think that there are many different Azure options at the moment, and it's a bit difficult to point users to a single canonical approach that they're most likely to have success with. I would like to simplify the first experience for Cilium users where possible.

One of my motivations to refer to this as "legacy" is that as I understand it used to be possible to use with AKS, and then AKS locked down the APIs to remove the ability to use this option. Given that many users who are interested on running Cilium on Azure will go directly to AKS, I would rather guide those users away from this particular option if it's going to cause issues for them. If it's a matter of just clearly documenting that this Azure IPAM option is for self-managed Azure environments, then I think that this would satisfy my concern there.

The other broader question is around the support of this code. As long as we have community members like @hemanthmalla who are happy to keep that code maintained, I think it is reasonable to keep within Cilium. Some background here is that Azure are free to change their IPAM implementations over time, and broadly in the Cilium community we aren't following very closely how that works. Upstream can make changes to their cloud + first-party IPAM plugin + the libraries at any time in a way that Cilium would then have to catch up on in this code to rebase + adjust to suit. As a general rule, the newer model with "delegated IPAM" is simpler for Cilium code maintenance because there is a clear API that separates the responsibilities of IPAM vs. other CNI responsibilities, and if we use this then we don't need to maintain & update this code. Upstream seems to be providing a canonical well-maintained version of the plugin which we just need to consume. At the same time, as I understand some of the motivation to support and use the Cilium Azure IPAM plugin here is that it is more efficient with the cloud API access because we rely on the single operator in the cluster to interact with cloud APIs and relay that information consistently within the cluster. From what I understand, this is not necessarily true with the upstream plugin. From that perspective, it sounds like something that is possible today with Cilium would not be possible with other models. This reinforces my opinion that at this time, as long as those use cases are not being adequately solved by other projects, and as long as we have Cilium community members willing to maintain this support, we should keep it around.

So then from my perspective it mainly comes down to exactly how we phrase the docs to highlight the paths that will make most users as successful as we can. "Legacy" wording helps with this by steering folks away from this option, but I think that it's also a loaded term that implies it will be going away some time soon. I believe that this is part of the point of contention here. We don't have a plan to remove it at this time, but if users do want to use it then we expect those users to be more savvy regarding the choice of configuration, and as always we would welcome additional help with maintaining this Azure IPAM code.

**Comment by hemanthmalla** [comment]
Thanks for the excellent summary @joestringer 

> AKS locked down the APIs to remove the ability to use this option

I'm really curious why Azure would do this, IIUC this would force users to have their nodes configured with elevated privileges to modify the VM. In addition to that users would loose centralized visibility and ability to enforce rate limits. @tamilmani1989 do you happen to know any reasoning behind this ?

The other aspect is having standardized abstractions to deal with across cloud providers ( pre-allocate, min-allocate etc). From a UX perspective, users cannot do away with worrying about their IP pools and now they'll need to worry about two CNI plugins. I understand the argument about ease of maintenance though.  We'll definitely continue to help with maintaining the Azure CNI mode.

If the intention is only to help AKS users pick the correct mode, we could remove any references of AKS in [Azure IPAM documentation](https://docs.cilium.io/en/stable/network/concepts/ipam/azure/#ipam-azure) and maybe rename `azure` ipam mode to `azure-non-aks` ?

**Comment by tamilmani1989** [comment]
Regarding this 
> AKS locked down the APIs to remove the ability to use this option
Cilium with azure-ipam is never supported by AKS officially. If I'm correct, I think AKS cluster was created with azure cni configuration and then cilium replaces it with cilium cni and azure ipam.

@hemanthmalla The main concern that joe brought out is that supportability of azure ipam scenario with cilium. I see your point though for self managed k8s clusters. One other option is to run cilium in tunnel mode but if requirement is direct routing with vnet ips, then there is no option today. Just to give background, initially AKS with Azure CNI(legacy) also used similar approach as like azure ipam but we noticed potential scalability issues with allocating IPs from azure network stack via that route and moved to new and better path today. 

Being said that, if required, I can volunteer to help from Azure side (for this configuration) until we have good story for self managed Kubernetes clusters.

**Comment by tamilmani1989** [comment]
Removed `legacy` and updated recommended ways to install cilium on AKS

**Review by gandro — APPROVED** [review]
Thanks!

**Review by tommyp1ckles — APPROVED** [review]

File: Documentation/network/concepts/ipam/azure.rst [hunk 1]
```diff
@@ -14,7 +14,10 @@ Azure IPAM
 
    While still maintained for now, Azure IPAM is considered legacy and is not
    compatible with AKS clusters created in `Bring your own CNI <https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli>`_
-   mode. Using BYOCNI is the preferred way to install Cilium on AKS.
+   mode. The recommended way to install cilium on AKS are 
+   `Bring your own CNI <https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli>`_ or
+   `Azure CNI Powered by Cilium <https://aka.ms/aks/cilium-dataplane>`_.
```
Comment (learnitall) [inline Documentation/network/concepts/ipam/azure.rst]:
```suggestion
   `Bring your own CNI <https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli>`__ or
   `Azure CNI Powered by Cilium <https://aka.ms/aks/cilium-dataplane>`__.
```

Can you add an extra underscore at the end of these links? That should fix the linting issues in the Documentation workflow.

**Review by learnitall — DISMISSED** [review]

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
The changes in this PR LGTM and are straightforward. I see that the text still refers to the embedded Azure IPAM in Cilium as "legacy" but perhaps going forward we should use other phrasing like "classic". That said, we can follow up on that separately, no need to further block this PR.

**Pull request merged**